polly/src/event_manager.rs

```diff
@@ -0,0 +1,786 @@
+/// using callbacks.
+pub trait EventHandler: Send {
+    /// Handle a read event (EPOLLIN).
+    fn handle_read(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
+        None
+    }
+    /// Handle a write event (EPOLLOUT).
+    fn handle_write(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
+        None
+    }
+    /// Handle a close event (EPOLLRDHUP).
+    fn handle_close(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
+        None
+    }
+    /// Handle an error event (EPOLLERR).assert_ne!
+    fn handle_error(&mut self, _source: Pollable) -> Option<Vec<PollableOp>> {
+        None
+    }
+
+    /// Initial registration of pollable objects.
+    /// Use the PollableOpBuilder to build the vector of PollableOps.
+    fn init(&self) -> Option<Vec<PollableOp>>;
+}
+
+/// Wraps a HashMap of fd to EventHandlerData.
+struct HandlerMap {
+    handlers: HashMap<i32, EventHandlerData>,
+}
+
+impl HandlerMap {
+    pub fn new() -> HandlerMap {
+        HandlerMap {
+            handlers: HashMap::new(),
+        }
+    }
+
+    pub fn get_mut_ref(&mut self, id: i32) -> Option<&mut EventHandlerData> {
+        self.handlers.get_mut(&id)
+    }
+
+    // Returns a copy of EventHandlerData instead of mut ref.
+    pub fn get(&mut self, id: i32) -> Option<EventHandlerData> {
+        match self.handlers.get(&id) {
+            Some(handler) => Some(EventHandlerData::new(
+                (handler.data.0.clone(), handler.data.1),
+                handler.handler.clone(),
+            )),
+            None => None,
+        }
+    }
+}
+
+impl Deref for HandlerMap {
+    type Target = HashMap<i32, EventHandlerData>;
+    fn deref(&self) -> &Self::Target {
+        &self.handlers
+    }
+}
+
+impl DerefMut for HandlerMap {
+    fn deref_mut(&mut self) -> &mut HashMap<i32, EventHandlerData> {
+        &mut self.handlers
+    }
+}
+
+// The tx side of a generic API channel.
+// Sending a T message over this channel will also write the eventFd.
+// The rx side of the channel polls the eventfd for incoming messages.
+pub struct GenericApiChannel<T> {
+    channel: Sender<T>,
+    fd: EventFd,
+}
+
+impl<T> GenericApiChannel<T> {
+    pub fn new(channel: Sender<T>) -> GenericApiChannel<T> {
+        GenericApiChannel {
+            channel,
+            fd: EventFd::new().unwrap(),
+        }
+    }
+
+    /// Send a message of type T and notify rx side by writing
+    /// the eventfd.
+    pub fn send(&mut self, msg: T) -> Result<()> {
+        self.fd.write(1).map_err(Error::ApiChannelFd)?;
+        self.channel.send(msg).map_err(|_| Error::ApiChannelTx)
+    }
+
+    /// Reads evenfd event count.
+    pub fn read_event(&mut self) -> u64 {
+        self.fd.read().unwrap_or(0)
+    }
+}
+
+impl<T> Clone for GenericApiChannel<T> {
+    fn clone(&self) -> Self {
+        GenericApiChannel {
+            channel: self.channel.clone(),
+            fd: self.fd.try_clone().unwrap(),
```

Comment:
It might be unlikely, but this unwrap can panic when `dup`call on the fd fails. Panic in the implementation of clone is usually unexpected and should be avoided. Can we work around implementing clone for `GenericApiChannel`? The first thing I can think of is to have a `try_clone()` function in `GenericApiChannel` and propagate the error from `self.fd.try_clone()`. This is not necessarily nice, and I would rather avoid that as well.